[2번째 시도] 웹 컴포넌트로 다시 구현한 자판기

live-small opened the pull request:
## 구현 결과물
[2번째로 만든 자판기](https://live-small.github.io/javascript-vendingmachine/)

### 구현한 기능 
- [x] `상품관리`페이지에서, `추가하기` 버튼을 통해 자판기에 상품을 추가할 수 있다. 
- [x] `상품관리`페이지에서, `삭제하기` 버튼을 통해 추가된 상품을 삭제할 수 있다. 
- [x] `잔돈충전`페이지에서, `충전하기` 버튼을 통해 자판기에 돈을 충전할 수 있다. 
- [x] `상품구매`페이지에서, `투입하기` 버튼을 통해 자판기에 돈을 투입할 수 있다. 
- [x] `상품구매`페이지에서, `구매하기` 버튼을 통해 해당 상품을 구입할 수 있다.
- [x] `상품구매`페이지에서, `반환하기` 버튼을 통해 투입한 금액을 반환할 수 있다.

## 구조 
![설계](https://user-images.githubusercontent.com/70274947/165695306-393e6313-e04f-4828-9ace-643611d428c2.png)


- 최상단 컴포넌트인 Compoent는 리액트 컴포넌트처럼 useState(setState -> render)를 구현해서, `상품관리, 잔돈충전, 상품구매`컴포넌트에서 이벤트 발생 시 이용할 수 있도록 했다. 

- UI랑 로직을 분리하고 싶어서, 
  - UI랑 이벤트 연결은 각 컴포넌트에서 수행
  - 컴포넌트에서 필요한 객체(Product, UserCoin, VendingMachineCoin)를 주입받아서 이벤트 핸들러에서 이용하도록 설계했다. 

## 고민한 부분 
### 이벤트 핸들링 
 [상품구매]에서 상품마다 구매하기 버튼이 있다. 이때 구매하기 버튼마다 이벤트 핸들러를 달아주는 게 비효율적이라고 생각해서, `구매할 수 있는 상품목록` 테이블에 이벤트 핸들러를 달아, event.target으로 클릭된 상품을 찾는 `이벤트위임`방식으로 구현했다. 

### 데이터 형태 
 데이터 불변성을 지키고 싶어 필요한 데이터 형태를 객체로 작업했다. 현재는`상품관리`에서 `product`를 추가하는 작업밖에 없어서, `{name, price, quantity}`형태를 원소로 가지는 배열로 작업했는데, 만약 추가된 `product`를 수정하거나 삭제한다고 하면 모든 상품목록을 돌아야하는 게 비효율적일 거 같았다. 그래서, 상품명을 key로 가지는 객체형태로 작업하려고 하니 동적으로 결정되는 key를 구조분해할당으로 받아올 수가 없었다. 이 pr을 쓰면서 든 생각인데, `dataset`속성을 이용해, 배열 인덱스를 이용하면 되지않을까? 

추가로 자잘한 고민은 코드에 직접 표시해두겠습니다 : )

live-small committed:
docs: 구현할 기능목록과 구현 전 가정 작성

live-small committed:
etc: .gitignore에 vscode 설정파일 추가

live-small committed:
style: eslint, prettier 설정

live-small committed:
feat: 첫 화면은 상품관리, 잔돈충전, 상품 구매 버튼만 보이도록 구현

live-small committed:
feat: 상품관리 버튼 클릭하면 상품추가폼, 상품현황 보이도록 구현

live-small committed:
feat: 잔돈충전 버튼 클릭하면 잔돈충전폼, 보유금액, 보유한 동전 개수가 보이도록 구현

live-small committed:
feat: 상품구매 버튼 클릭시 보여지는 화면 구현

보여지는 화면: 금액 투입하는 폼, 투입한 금액, 구매할 수 있는 상품현황, 반환된 잔돈 개수

live-small committed:
feat: [상품관리] 추가하기 버튼으로 상품을 추가할 수 있다

live-small committed:
feat: [상품관리] 한번 추가한 상품은 새로고침해도 유지된다

로컬 스토리지로 데이터 관리

live-small committed:
feat: [잔돈충전] 충전하기 버튼으로 입력한 돈을 충전할 수 있다

live-small committed:
feat: [상품구매] 상품관리에서 등록한 상품을 구매할 수 있는 상품현황에서 볼 수 있다

live-small committed:
docs: 상품구매탭에서, 구매하기 이벤트 기능목록에 추가

live-small committed:
fix: 등록된 상품, 자판기가 보유한 동전 값은 [상품구매]에서도 이용하기 때문에 부모 클래스(component)에 getter 설정

live-small committed:
feat: [상품구매] 투입하기 버튼 클릭시, 자판기에 돈을 투입할 수 있다

live-small committed:
[상품구매] 상품 옆 구매하기 버튼을 클릭해, 상품을 구매할 수 있다

live-small committed:
refactor: 각각의 컴포넌트에서 setState 메서드 이용 -> 최상위 클래스(Component)에서 setState method로 통일해 재사용

live-small committed:
refactor: 상품관리, 상품구매에서 product 객체를 주입받아 이용하도록 수정

최상단 컴포넌트(Component)에서 상품 정보 getter을 선언했는데, 리액트 컴포넌트처럼 useState(setState, render)이용하기 위해 만든 Component 클래스 의미 변질, 그래서 Product 객체를 만들어, 필요한 곳에 주입해서 이용하는 방법으로 수정

live-small committed:
refactor: 잔돈충전에서 vendingMachineCoin 객체를 주입받아 이용하도록 수정

live-small committed:
feat: [상품구매] 구매하기버튼으로 상품 구매시, 투입한 돈과 자판기 보유동전이 업데이트된다

사용자는 상품을 얻게되고, 자판기는 상품가격을 얻게된다

live-small committed:
fix: 마크업 수정

live-small committed:
feat: [상품구매] 반환하기 버튼을 누르면, 투입한 금액의 최소 개수의 동전으로 반환된다

live-small committed:
fix: [상품구매] 마크업 태그속성 정해준대로 수정 (class -> id)

live-small committed:
fix: [상품관리] 상품 추가 후 업데이트할 때 필요한 키, 주입받은 product 프로퍼티 이용

live-small committed:
fix: 테스트 코드보고, 요구한 기능대로 수정

기능 요구사항에 정확히 기재되어있지 않아, 자판기에서 상품을 구매하면서 발생하는 돈을 자판기 보유동전에 추가하도록 구현했다. 하지만 테스트 코드를 보니 그렇지 않아서 수정했다

live-small committed:
feat: [잔돈충전] 충전할 돈에 음수값이 올 경우 에러처리 추가

live-small committed:
refactor: 불필요한 렌더링을 막기위해, local storage에 값을 변경한 후 마지막에 렌더링하도록 수정

live-small committed:
refactor: local storage 에서 값 얻기, 설정, 초기화함수 utils > localstorage.js로 분리해서 재사용

live-small committed:
etc: 불필요한 주석 제거

live-small commented on src/components/VendingMachineCoin.js:
```diff
@@ -0,0 +1,73 @@
+    constructor() {
+        this.key = "vendingMachineCoin";
+        this.coinTemplate = { 500: 0, 100: 0, 50: 0, 10: 0 };
+        initLocalStorage(this.key, {
+            totalCoin: 0,
+            numberOfCoin: { ...this.coinTemplate },
+        });
```
자판기가 보유한 동전 값과 동전 개수 정보가 담긴 numberOfCoin을 지금처럼 같은 객체로 관리하는 게 좋을지, 각각 따로 관리하는 게 좋을지 고민이 됩니다. 둘 다 자판기가 가진 동전의 정보이고, 표현방식만 다르다고 생각해서 하나의 객체로 관리했습니다!

JODAHYE replied:
저는 문자열로 하나씩 저장했었는데 ㅜㅠ 객체로 값을 저장하니 더 직관적이고 좋아보이네욥!!

live-small replied:
리뷰 감사합니다. @JODAHYE  
totalCoin(총 동전값)과 numberOfCoin(동전개수정보)은 둘 다 자판기가 가진 동전의 정보이니까 같은 객체로 관리하는거랑 각각 따로 하는 거에 대해선 어떻게 생각하시나요?

live-small commented on src/template/CoinCharge.js:
```diff
@@ -0,0 +1,38 @@
+	<br />
+	<div class="coin-current-status">
+		<h2>자판기가 보유한 동전</h2>
+		<table>
+			<tr>
+				<th>동전</th>
+				<th>개수</th>
+			</tr>
+			<tr>
+				<td>500원</td>
+				<td id="vending-machine-coin-500-quantity">${numberOfCoin[500]}개</td>
+			</tr>
+			<tr>
```
html tag 간격이 이상한 거 같네요

JODAHYE replied:
탭 간격은 eslint 설정때문인가요?!

live-small replied:
탭 간격은 prettier로 설정했습니다. 
제 로컬에서는 prettier에 tabwidth 4로 잘 저장되어있는데, 깃허브 상에선 tabwidth가 8로 되어있네요. 이유찾고 수정하겠습니다. 

제 깃헙 세팅에 tab width가 8(default)로 설정되어있어서, 이 파일처럼 간격이 8로 렌더링 됐네요. 4로 바꾸니까 자동으로 바꼈습니다 : ) 
(settings > Appearance > tab width) 
[#참고](https://stackoverflow.com/questions/8833953/how-to-change-tab-size-on-github)

live-small commented on src/main.js:
```diff
@@ -0,0 +1,35 @@
+/* eslint-disable no-new */
```
eslint에서 new 키워드를 이용하면 에러가 나는데, 어떤 이유에서 에러나도록 설정해뒀는지 찾아봅시다!

live-small commented on src/components/VendingMachineCoin.js:
```diff
@@ -0,0 +1,73 @@
+    return(coinToReturn) {
+        const returnNumberOfCoin = { ...this.coinTemplate };
+        const haveNumberOfCoin = { ...this.data.numberOfCoin };
+        const coinKinds = Object.keys(haveNumberOfCoin).reverse();
+        for (const unit of coinKinds) {
```
haveNumberOfCoin이 `{500: 0, 100: 0, 50:0, 10:0 }` 으로 로컬스토리지에 설정했는데, 로컬 스토리지에서 받아올 땐 `{10:0, 50:0 ~~ }` 으로 되어있어서, reverse처리를 해줬는데 왜 거꾸로 설정되는지 이유찾고, 수정해야겠습니다.

live-small replied:
### 이유 
javascript의 object는 **순서를 보장하지 않는 자료구조**였습니다. (객체는 늘 key로 검색하다보니까 잊고있었네요..)
그래서, **Map 자료구조**가 이 단점을 보완했습니다. 
다만 아쉬운건, Map 자료구조는 `set(key,value)`로 하나씩 설정하거나, `[[key, value], [key2, value2]]` 이런 형태로 인자에 넘겨줘야 하더라구요. 
현재는 key, value가 4개여서 하나씩 적어도 되는데, 더 많아질 수도 있으니까 map을 설정하는 함수를 만드는 것도 생각해봐야겠습니다. 

#### Object.freeze
사실, `Object.freeze`써서 얼리면 되지 않을까 했는데! 
Object.freeze는 객체 생성이후, 동적으로 객체 추가, 삭제를 방지하는 역할을 할 뿐, 객체의 순서보장과는 관계가 없다는 걸 알게되었습니다. 

#

참고 
[Does JavaScript Guarantee Object Property Order?](https://javascript.plainenglish.io/does-javascript-guarantee-object-property-order-36c3630e320b)
[Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)
[Object.freeze](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze)

live-small commented on src/components/VendingMachineCoin.js:
```diff
@@ -0,0 +1,73 @@
+    generateCoinRandomly(coinToInput) {
+        const coinTemplate = { ...this.coinTemplate };
+        while (coinToInput > 0) {
+            const coinKinds = Object.keys(coinTemplate).map(unit => Number(unit));
```
10원을 입력했을 때 투입한 돈만큼 랜덤으로 동전이 생긴다해도 10원만 생성될텐데,
무작위로 동전이 생기기 때문에 while문은 1번만 돌지 않습니다. 연산량을 보장하려면 어떻게 수정할 수 있을까요?

JODAHYE replied:
음 투입한 돈의 범위에 따라 생성될 수 있는 동전 단위를 배열로 반환하는 함수를 따로 만들어서 const coinkinds에 담는거 어떨까요
40원 이하면 [10] 
50원이상 90원 이하면 [10, 50],
100원이상 490원 이하면 [10, 50, 100],
 500이상이면[10, 50, 100, 500]

live-small replied:
@JODAHYE 오 그런방법도 있네요 !! 좋은 방법 감사합니다! 리팩토링할 때 반영해보겠습니다.

live-small commented on src/components/ProductPurchase.js:
```diff
@@ -0,0 +1,85 @@
+        });
+    }
+
+    checkInsertCoin(needToCoin) {
+        if (this.UserCoin.InsertCoin - needToCoin < 0) {
+            return alert(
+                `돈이 부족합니다. ${Math.abs(
+                    this.UserCoin.InsertCoin - needToCoin
+                )}원 더 투입해주세요`
+            );
+        }
+        return true;
+    }
```
이 함수가 여기 위치하는 게 적절할지, validator로 옮겨서 필요한 값을 파라미터로 넘겨주는 게 나을지 고민이 들었는데, 후자가 좋을 거 같네요

JODAHYE replied:
한 파일에서만 써서 여기 위치 시킨건가용?

live-small replied:
네 그리고, ProductPurchase 클래스에 주입받는 UserCoin의 InsertCoin을 이용하기 위해서 여기 위치시켰습니다. 
그런데 UI를 그리고, 이벤트 핸들러만 달아주는 컴포넌트에 위치하는 건 적절하지 않은 거 같네요. 
validator 폴더로 이동하도록 수정하겠습니당 : )

live-small commented on src/components/VendingMachineCoin.js:
```diff
@@ -0,0 +1,73 @@
+        const returnNumberOfCoin = { ...this.coinTemplate };
+        const haveNumberOfCoin = { ...this.data.numberOfCoin };
+        const coinKinds = Object.keys(haveNumberOfCoin).reverse();
+        for (const unit of coinKinds) {
+            if (!coinToReturn) break;
+            const needCoinCount = Math.floor(coinToReturn / unit);
+            const count = Math.min(needCoinCount, haveNumberOfCoin[unit]);
+            coinToReturn -= unit * count;
+            returnNumberOfCoin[unit] += count;
+            haveNumberOfCoin[unit] -= count;
+        }
```
최소 개수의 동전을 반환하기위해서, 현재 동전 개수 중 높은 단위 동전부터(500, 100, 50, 10)돌면서 빼도 되지만 
**반환해야할 돈이 엄청 클 경우(5만원) 연산하는 데 시간이 많이 들 거 같아서, 어떤 수든 연산량 보장이 가능하도록 구현하고 싶었습니다.** 그래서 사용하는 동전 종류를 돌면서, 현재 가진 동전 개수, 투입한 돈에서 필요한 동전 개수의 최소값을 이용했습니다. 

예를 들어, 만약 2200원을 반환해야할 때 
가진 개수: 500원 3개
필요한 개수: 500원 4개 
: 3개만 사용 

가진 개수: 500원 5개
필요한 개수: 500원 4개 
: 4개만 사용

JODAHYE replied:
오!! 좋은 방법이네요 이런 방법이 있었다니

live-small reviewed (commented):
고민한 부분 & 1차 셀프리뷰

JODAHYE commented on src/utils/utils.js:
```diff
@@ -0,0 +1 @@
+export const $ = selector => document.querySelector(selector);
```
오 빠르게 element를 찾을 수 있게 하셨네요! 굿

JODAHYE commented on src/components/Product.js:
```diff
@@ -0,0 +1,32 @@
+        }
+        return productList;
+    }
+}
```
오 상품들을 객체배열로 관리하신건가요?

live-small replied:
네! 상품에서 다루는 데이터 name, price, quantity를 속성으로 가지는 객체를 배열로 관리했습니다!

JODAHYE commented on src/components/ProductManage.js:
```diff
@@ -0,0 +1,29 @@
+            }
+        });
+    }
+}
```
올바른 값을 입력하도록 검증하는 코드를 함수로 분리하니 간결해보이고 좋네요

JODAHYE commented on src/main.js:
```diff
@@ -0,0 +1,35 @@
+    }
+}
+
+new VendingMachine($("#app")).render();
```
메인에서 탭메뉴별로 렌더링하도록 하는 로직만 있으니 깔끔하고 보기 좋아요

JODAHYE reviewed (commented):
코드 보니까 깔끔하고 잘 짜신 것 같아요 제가 얼마나 지저분하게 했는지 깨닫게 되고 이런 방법도 있구나 하고 깨닫는 것도 많고 도움 많이 되었습니다!
더불어 클래스의 장점도 알게 되었네용 !!

live-small commented on .gitignore:
```diff
@@ -1 +1,2 @@
 node_modules
+.vscode
```
end of line 처리해야합니당

live-small committed:
fix: gitignore 파일에 EOL(end of line)추가

live-small committed:
fix: [상품구매] 컴포넌트에선 UI, 이벤트 바인딩만 관리하기 위해 유효성 체크하는 함수 validator로 이동

자판기에 돈이 없어 잔돈을 다 반환하지 못할 때, 반환가능한 가격만이라도 반환하도록 로직 수정

live-small committed:
fix: 자판기 이용을 쉽게하기 위해 input에 placeholder 속성 추가

live-small committed:
docs: 추가작업 목록 추가

live-small committed:
refactor: [잔돈충전] 투입한 돈만큼 랜덤으로 동전 생성할 때, 랜덤 생성할 동전종류를 투입된 돈에 따라 다르게 처리해 연산량 줄임

40원일 땐, 10,50,100,500원에서 10원만 나오기위해 4번 이상을 돌았지만, 50원보다 작을 경우, 10원만 랜덤생성 모듈에 넣어서 4번으로 연산량 보장

live-small committed:
fix: 충전된 돈으로 랜덤생성할 때, 이용할 동전종류찾는 함수(getCoinKindsToUse)에서 투입된 돈과 동전 종류가 같을 때도 해당 동전 종류 이용하는 케이스 처리

10원 투입됐을 때, 10원 동전 종류도 이용할 수 있다

live-small committed:
feat: [상품관리] 제품 옆 삭제하기 버튼을 통해 상품을 삭제할 수 있다

위험을 방지하고자 삭제 전, 삭제하시겠습니까? 창에 확인을 누른 경우만 삭제진행

live-small commented:
### 변경된 사항 
- [투입된 돈만큼 동전 랜덤생성하는 함수, 연산량 줄임](https://github.com/live-small/javascript-vendingmachine/pull/2/commits/7a13ba7e0e912d63f182a643d55ef4fe57f3b71b)
- [상품관리에서 추가된 상품 삭제 기능 추가](https://github.com/live-small/javascript-vendingmachine/pull/2/commits/77cd557a3e91e85d8f74446d69f65e6df25936ed)

### 현재 고민 
- 자판기에 이용하는 동전 객체 `{500: 0, 100:0, 50:0, 10:0}`를 Map 자료구조로 변경해, key의 순서를 보장하려고 하는데 동적으로 추가되지 않게끔 고정하려면(Object.freeze) 어떻게 해야할까?
- 상품관리에서 이미 추가된 상품은 추가안되도록 구현하려하는데(재고관리할 때, 헷갈릴 수 있으니까 상품수정기능으로 이름, 가격, 수량 변경하도록 하려고 함), 현재 상품객체배열 전체를 돌면서 체크해야하는데 상품을 추가할 때마다 상품 전체를 도니까 비효율적일 거 같다. 
```javascript
const productList: [{
    coke: {
        name: coke;
        price: 1500;
        quantity: 30;
    };
}]
```
상품을 이런 형태로 관리한다면, **동적으로 정해지는 키를 가진 객체들의 배열을 만들 때, map으로 객체를 돈다면 어떻게 키에 접근하지?** 
- 상품관리에서 추가된 상품을 수정하려면 어떻게 구현하지?

live-small commented:
## 고민중인 부분
### 상품 데이터 형식을 어떻게 하는게 좋을까?  

  ```javascript
// 생각한 데이터 형식 
        const product = {
            coke: { price: 1500, quantity: 30 },
            sprite: { price: 1500, quantity: 20 },
        };

// 동적으로 key를 설정하려면 이렇게! 
     const newProduct = { [name]: { price, quantity } };
```
    
#### 고려해야할 부분
  1. 중복된 상품인지 체크
            모든 상품리스트를 돌면서 상품명을 체크하는 것보단, 상품명을 key로 가지는 객체의 배열로 하는 건 어떨까?
            상품명으로 접근했을 때 값이 없으면, 추가되지 않은 상품인 걸 1번에 알 수 있다.  

  2. 상품 삭제
            현재 배열로 관리했기에 product-key로 splice해서 삭제했는데, 객체로 된다면 delete 키워드로 삭제할 수 있다.
            근데 자바스크립트 객체는 순서를 보장하지 않기때문에 렌더링했을 때 상품 목록이 달라질 수 있다.
            순서를 보장하는 Map 자료구조로 바꾸고, delete method를 쓰는 건 어떨까?
            **Map이 JSON변환 시, 객체로 맵핑 돼, 로컬 스토리지에서 받아올 때 Map 객체가 아님**
            객체로한다면, 어떻게 순서를 보장할 수 있을지 생각해봐야한다 \*\*
     
   3. 상품 수정 
          `상품구매`탭에서 상품을 구매할 경우, 상품 정보의 수량을 수정해야한다. 배열로 했을 땐 product-key로 splice하고, 수정한 상품을 추가해줬는데, 객체로 했을 땐 어떻게 해야할까..? 

#### product-data-object로 브랜치 파서 고민해보고 작업하기로 결정 !
